In [1]:
import pandas as pd

In [2]:
import numpy as np
import seaborn as sns
from pydataset import data

# One Variable in Multiple Columns

In [3]:
treatments = pd.read_csv('untidy-data/treatments.csv')
treatments

treatmenta  treatmentb  treatmentc
0    John Smith         NaN           2           0
1      Jane Doe        16.0          11           3
2  Mary Johnson         3.0           1           4

In [4]:
treatments.columns = ['name', 'a', 'b', 'c']
# we should rename columns for readability

treatments

name     a   b  c
0    John Smith   NaN   2  0
1      Jane Doe  16.0  11  3
2  Mary Johnson   3.0   1  4

In [5]:
treatments.melt(id_vars=['name'], var_name='treatment', value_name='response')

# .melt() will combine multiple columns into just 2 columns

# id_vars=[''] will identify columns that will not be melted

# var_name='' will rename the column that was combined

# value_name='' will rename the column containing the resulting values

## it is up to the practicioner to understand how the data has changed 

name treatment  response
0    John Smith         a       NaN
1      Jane Doe         a      16.0
2  Mary Johnson         a       3.0
3    John Smith         b       2.0
4      Jane Doe         b      11.0
5  Mary Johnson         b       1.0
6    John Smith         c       0.0
7      Jane Doe         c       3.0
8  Mary Johnson         c       4.0

# One Column with Multiple Variables

In [6]:
students = pd.read_csv('untidy-data/students.csv')
students.head(9)

date                 var          val
0  2019-02-04   n_late_from_break     4.028120
1  2019-02-04  coffee_consumption  5255.409740
2  2019-02-04      classroom_temp    67.000000
3  2019-02-05   n_late_from_break     2.101998
4  2019-02-05  coffee_consumption  8603.704719
5  2019-02-05      classroom_temp    73.000000
6  2019-02-06   n_late_from_break     4.941244
7  2019-02-06  coffee_consumption  1801.498050
8  2019-02-06      classroom_temp    81.000000

In [7]:
students.pivot(index='date', columns='var')

val                                     
var        classroom_temp coffee_consumption n_late_from_break
date                                                          
2019-02-04           67.0        5255.409740          4.028120
2019-02-05           73.0        8603.704719          2.101998
2019-02-06           81.0        1801.498050          4.941244
2019-02-07           62.0        9282.959741          1.419342
2019-02-08           72.0        7558.270659          1.808919
2019-02-11           60.0        5731.008713          4.831584
2019-02-12           52.0        9547.673484          3.550830
2019-02-13           56.0        7114.931847          2.550362
2019-02-14           72.0        1977.295513          3.507548
2019-02-15           75.0        2254.674679          2.114379
2019-02-18           55.0        4819.981878          2.249609
2019-02-19           82.0        8628.442171          4.646607
2019-02-20           58.0        6905.019187          1.309552
2019-02-21           79.0        1414.705565          2.132542
2019-02-22           68.0        8894.846285          1.745758
2019-02-25           84.0        5982.280540          2.344274
2019-02-26           54.0        9121.391375          4.620919
2019-02-27           88.0        4412.125903          1.860681
2019-02-28           81.0        1693.233409          1.342123

In [8]:
students.set_index(['date', 'var']).unstack(level=1)

val                                     
var        classroom_temp coffee_consumption n_late_from_break
date                                                          
2019-02-04           67.0        5255.409740          4.028120
2019-02-05           73.0        8603.704719          2.101998
2019-02-06           81.0        1801.498050          4.941244
2019-02-07           62.0        9282.959741          1.419342
2019-02-08           72.0        7558.270659          1.808919
2019-02-11           60.0        5731.008713          4.831584
2019-02-12           52.0        9547.673484          3.550830
2019-02-13           56.0        7114.931847          2.550362
2019-02-14           72.0        1977.295513          3.507548
2019-02-15           75.0        2254.674679          2.114379
2019-02-18           55.0        4819.981878          2.249609
2019-02-19           82.0        8628.442171          4.646607
2019-02-20           58.0        6905.019187          1.309552
2019-02-21           79.0        1414.705565          2.132542
2019-02-22           68.0        8894.846285          1.745758
2019-02-25           84.0        5982.280540          2.344274
2019-02-26           54.0        9121.391375          4.620919
2019-02-27           88.0        4412.125903          1.860681
2019-02-28           81.0        1693.233409          1.342123

In [9]:
tidy_students = students.set_index(['date', 'var']).unstack(level=1)
tidy_students.columns = tidy_students.columns.droplevel()
tidy_students.columns.name = ''

tidy_students

classroom_temp  coffee_consumption  n_late_from_break
date                                                             
2019-02-04            67.0         5255.409740           4.028120
2019-02-05            73.0         8603.704719           2.101998
2019-02-06            81.0         1801.498050           4.941244
2019-02-07            62.0         9282.959741           1.419342
2019-02-08            72.0         7558.270659           1.808919
2019-02-11            60.0         5731.008713           4.831584
2019-02-12            52.0         9547.673484           3.550830
2019-02-13            56.0         7114.931847           2.550362
2019-02-14            72.0         1977.295513           3.507548
2019-02-15            75.0         2254.674679           2.114379
2019-02-18            55.0         4819.981878           2.249609
2019-02-19            82.0         8628.442171           4.646607
2019-02-20            58.0         6905.019187           1.309552
2019-02-21            79.0         1414.705565           2.132542
2019-02-22            68.0         8894.846285           1.745758
2019-02-25            84.0         5982.280540           2.344274
2019-02-26            54.0         9121.391375           4.620919
2019-02-27            88.0         4412.125903           1.860681
2019-02-28            81.0         1693.233409           1.342123

# A More Complex Example

In [10]:
sales = pd.read_csv('./untidy-data/sales.csv')
sales

Product  2016 Sales  2016 PPU  2017 Sales  2017 PPU  2018 Sales  2018 PPU
0       A         673         5         231         7         173         9
1       B         259         3         748         5         186         8
2       C         644         3         863         5         632         5
3       D         508         9         356        11         347        14

In [11]:
sales_melt = sales.melt(id_vars='Product', var_name='year_and_measure')
sales_melt.head()

Product year_and_measure  value
0       A       2016 Sales    673
1       B       2016 Sales    259
2       C       2016 Sales    644
3       D       2016 Sales    508
4       A         2016 PPU      5

In [12]:
year_and_measure_df = sales_melt.year_and_measure.str.split(' ', expand=True)
year_and_measure_df.columns = ['year', 'measure']
year_and_measure_df.head()

year measure
0  2016   Sales
1  2016   Sales
2  2016   Sales
3  2016   Sales
4  2016     PPU

In [13]:
sales2 = pd.concat([sales_melt, year_and_measure_df], axis=1).drop(columns='year_and_measure')
sales2.head()

Product  value  year measure
0       A    673  2016   Sales
1       B    259  2016   Sales
2       C    644  2016   Sales
3       D    508  2016   Sales
4       A      5  2016     PPU

In [14]:
sales_tidy = sales2.pivot_table(index=['Product', 'year'], columns='measure', values='value')
sales_tidy

measure       PPU  Sales
Product year            
A       2016    5    673
        2017    7    231
        2018    9    173
B       2016    3    259
        2017    5    748
        2018    8    186
C       2016    3    644
        2017    5    863
        2018    5    632
D       2016    9    508
        2017   11    356
        2018   14    347

In [15]:
sales_tidy.columns.name = ''
sales_tidy.reset_index(inplace=True)
sales_tidy

Product  year  PPU  Sales
0        A  2016    5    673
1        A  2017    7    231
2        A  2018    9    173
3        B  2016    3    259
4        B  2017    5    748
5        B  2018    8    186
6        C  2016    3    644
7        C  2017    5    863
8        C  2018    5    632
9        D  2016    9    508
10       D  2017   11    356
11       D  2018   14    347

# Exercises

## 1. Attendance Data

## Load the attendance.csv file and calculate an attendnace percentage for each student. One half day is worth 50% of a full day, and 10 tardies is equal to one absence.

In [16]:
attendance = pd.read_csv('untidy-data/attendance.csv')

attendance

Unnamed: 0 2018-01-01 2018-01-02 2018-01-03 2018-01-04 2018-01-05  \
0      Sally          P          T          T          H          P   
1       Jane          A          P          T          T          T   
2      Billy          A          T          A          A          H   
3       John          P          T          H          P          P   

  2018-01-06 2018-01-07 2018-01-08  
0          A          T          T  
1          T          A          T  
2          T          P          T  
3          T          P          P

In [17]:
attendance = attendance.rename(columns={'Unnamed: 0': 'student'})

In [18]:
attendance

student 2018-01-01 2018-01-02 2018-01-03 2018-01-04 2018-01-05 2018-01-06  \
0   Sally          P          T          T          H          P          A   
1    Jane          A          P          T          T          T          T   
2   Billy          A          T          A          A          H          T   
3    John          P          T          H          P          P          T   

  2018-01-07 2018-01-08  
0          T          T  
1          A          T  
2          P          T  
3          P          P

In [19]:
attendance = attendance.replace(['P', 'T', 'H', 'A'], [1, .9, .5, 0])

attendance

student  2018-01-01  2018-01-02  2018-01-03  2018-01-04  2018-01-05  \
0   Sally           1         0.9         0.9         0.5         1.0   
1    Jane           0         1.0         0.9         0.9         0.9   
2   Billy           0         0.9         0.0         0.0         0.5   
3    John           1         0.9         0.5         1.0         1.0   

   2018-01-06  2018-01-07  2018-01-08  
0         0.0         0.9         0.9  
1         0.9         0.0         0.9  
2         0.9         1.0         0.9  
3         0.9         1.0         1.0

In [20]:
attendance = attendance.melt(id_vars=['student'], var_name='date', value_name='absent')

attendance

student        date  absent
0    Sally  2018-01-01     1.0
1     Jane  2018-01-01     0.0
2    Billy  2018-01-01     0.0
3     John  2018-01-01     1.0
4    Sally  2018-01-02     0.9
5     Jane  2018-01-02     1.0
6    Billy  2018-01-02     0.9
7     John  2018-01-02     0.9
8    Sally  2018-01-03     0.9
9     Jane  2018-01-03     0.9
10   Billy  2018-01-03     0.0
11    John  2018-01-03     0.5
12   Sally  2018-01-04     0.5
13    Jane  2018-01-04     0.9
14   Billy  2018-01-04     0.0
15    John  2018-01-04     1.0
16   Sally  2018-01-05     1.0
17    Jane  2018-01-05     0.9
18   Billy  2018-01-05     0.5
19    John  2018-01-05     1.0
20   Sally  2018-01-06     0.0
21    Jane  2018-01-06     0.9
22   Billy  2018-01-06     0.9
23    John  2018-01-06     0.9
24   Sally  2018-01-07     0.9
25    Jane  2018-01-07     0.0
26   Billy  2018-01-07     1.0
27    John  2018-01-07     1.0
28   Sally  2018-01-08     0.9
29    Jane  2018-01-08     0.9
30   Billy  2018-01-08     0.9
31    John  2018-01-08     1.0

In [21]:
attendance.groupby('student').absent.mean()

student
Billy    0.5250
Jane     0.6875
John     0.9125
Sally    0.7625
Name: absent, dtype: float64

## 2. Coffee Levels

### Read the coffee_levels.csv file.

In [22]:
coffee_levels = pd.read_csv('untidy-data/coffee_levels.csv')

coffee_levels

hour coffee_carafe  coffee_amount
0      8             x       0.816164
1      9             x       0.451018
2     10             x       0.843279
3     11             x       0.335533
4     12             x       0.898291
5     13             x       0.310711
6     14             x       0.507288
7     15             x       0.215043
8     16             x       0.183891
9     17             x       0.391560
10     8             y       0.189297
11     9             y       0.521502
12    10             y       0.023163
13    11             y       0.235529
14    12             y       0.017009
15    13             y       0.997464
16    14             y       0.058361
17    15             y       0.144644
18    16             y       0.544676
19    17             y       0.594126
20     8             z       0.999264
21     9             z       0.915990
22    10             z       0.144928
23    11             z       0.311495
24    12             z       0.771947
25    13             z       0.398520
26    14             z       0.864464
27    15             z       0.436364
28    16             z       0.280621
29    17             z       0.436677

### Transform the data so that each carafe is in it's own column.

In [23]:
coffee_levels.pivot(index='hour', columns='coffee_carafe')

coffee_amount                    
coffee_carafe             x         y         z
hour                                           
8                  0.816164  0.189297  0.999264
9                  0.451018  0.521502  0.915990
10                 0.843279  0.023163  0.144928
11                 0.335533  0.235529  0.311495
12                 0.898291  0.017009  0.771947
13                 0.310711  0.997464  0.398520
14                 0.507288  0.058361  0.864464
15                 0.215043  0.144644  0.436364
16                 0.183891  0.544676  0.280621
17                 0.391560  0.594126  0.436677

### Is this the best shape for the data?

## 3. Cake Recipes

### Read the cake_recipes.csv data. This data set contains cake tastiness scores for combinations of different recipes, oven rack positions, and oven temperatures.

In [24]:
cake_recipes = pd.read_csv('untidy-data/cake_recipes.csv')

cake_recipes

recipe:position        225        250        275        300
0        a:bottom  61.738655  53.912627  74.414730  98.786784
1           a:top  51.709751  52.009735  68.576858  50.228470
2        b:bottom  57.095320  61.904369  61.196980  99.248541
3           b:top  82.455004  95.224151  98.594881  58.169349
4        c:bottom  96.470207  52.001358  92.893227  65.473084
5           c:top  71.306308  82.795477  92.098049  53.960273
6        d:bottom  52.799753  58.670419  51.747686  56.183110
7           d:top  96.873178  76.101363  59.571620  50.971626

### Tidy the data as necessary.

In [25]:
cake_recipes = cake_recipes.rename(columns={'recipe:position' : 'recipe_position'})

cake_recipes

recipe_position        225        250        275        300
0        a:bottom  61.738655  53.912627  74.414730  98.786784
1           a:top  51.709751  52.009735  68.576858  50.228470
2        b:bottom  57.095320  61.904369  61.196980  99.248541
3           b:top  82.455004  95.224151  98.594881  58.169349
4        c:bottom  96.470207  52.001358  92.893227  65.473084
5           c:top  71.306308  82.795477  92.098049  53.960273
6        d:bottom  52.799753  58.670419  51.747686  56.183110
7           d:top  96.873178  76.101363  59.571620  50.971626

In [26]:
cake_recipes = cake_recipes.melt(id_vars=['recipe_position'], var_name='temperature', value_name='tastiness')
    
cake_recipes

recipe_position temperature  tastiness
0         a:bottom         225  61.738655
1            a:top         225  51.709751
2         b:bottom         225  57.095320
3            b:top         225  82.455004
4         c:bottom         225  96.470207
5            c:top         225  71.306308
6         d:bottom         225  52.799753
7            d:top         225  96.873178
8         a:bottom         250  53.912627
9            a:top         250  52.009735
10        b:bottom         250  61.904369
11           b:top         250  95.224151
12        c:bottom         250  52.001358
13           c:top         250  82.795477
14        d:bottom         250  58.670419
15           d:top         250  76.101363
16        a:bottom         275  74.414730
17           a:top         275  68.576858
18        b:bottom         275  61.196980
19           b:top         275  98.594881
20        c:bottom         275  92.893227
21           c:top         275  92.098049
22        d:bottom         275  51.747686
23           d:top         275  59.571620
24        a:bottom         300  98.786784
25           a:top         300  50.228470
26        b:bottom         300  99.248541
27           b:top         300  58.169349
28        c:bottom         300  65.473084
29           c:top         300  53.960273
30        d:bottom         300  56.183110
31           d:top         300  50.971626

In [27]:
cake_recipes[['recipe', 'position']]= cake_recipes.recipe_position.str.split(':', expand = True)
    
cake_recipes

recipe_position temperature  tastiness recipe position
0         a:bottom         225  61.738655      a   bottom
1            a:top         225  51.709751      a      top
2         b:bottom         225  57.095320      b   bottom
3            b:top         225  82.455004      b      top
4         c:bottom         225  96.470207      c   bottom
5            c:top         225  71.306308      c      top
6         d:bottom         225  52.799753      d   bottom
7            d:top         225  96.873178      d      top
8         a:bottom         250  53.912627      a   bottom
9            a:top         250  52.009735      a      top
10        b:bottom         250  61.904369      b   bottom
11           b:top         250  95.224151      b      top
12        c:bottom         250  52.001358      c   bottom
13           c:top         250  82.795477      c      top
14        d:bottom         250  58.670419      d   bottom
15           d:top         250  76.101363      d      top
16        a:bottom         275  74.414730      a   bottom
17           a:top         275  68.576858      a      top
18        b:bottom         275  61.196980      b   bottom
19           b:top         275  98.594881      b      top
20        c:bottom         275  92.893227      c   bottom
21           c:top         275  92.098049      c      top
22        d:bottom         275  51.747686      d   bottom
23           d:top         275  59.571620      d      top
24        a:bottom         300  98.786784      a   bottom
25           a:top         300  50.228470      a      top
26        b:bottom         300  99.248541      b   bottom
27           b:top         300  58.169349      b      top
28        c:bottom         300  65.473084      c   bottom
29           c:top         300  53.960273      c      top
30        d:bottom         300  56.183110      d   bottom
31           d:top         300  50.971626      d      top

In [28]:
cake_recipes = cake_recipes.drop(columns='recipe_position')

cake_recipes

temperature  tastiness recipe position
0          225  61.738655      a   bottom
1          225  51.709751      a      top
2          225  57.095320      b   bottom
3          225  82.455004      b      top
4          225  96.470207      c   bottom
5          225  71.306308      c      top
6          225  52.799753      d   bottom
7          225  96.873178      d      top
8          250  53.912627      a   bottom
9          250  52.009735      a      top
10         250  61.904369      b   bottom
11         250  95.224151      b      top
12         250  52.001358      c   bottom
13         250  82.795477      c      top
14         250  58.670419      d   bottom
15         250  76.101363      d      top
16         275  74.414730      a   bottom
17         275  68.576858      a      top
18         275  61.196980      b   bottom
19         275  98.594881      b      top
20         275  92.893227      c   bottom
21         275  92.098049      c      top
22         275  51.747686      d   bottom
23         275  59.571620      d      top
24         300  98.786784      a   bottom
25         300  50.228470      a      top
26         300  99.248541      b   bottom
27         300  58.169349      b      top
28         300  65.473084      c   bottom
29         300  53.960273      c      top
30         300  56.183110      d   bottom
31         300  50.971626      d      top

In [41]:
cake_tidy = cake_recipes.pivot(index=['recipe', 'position'], columns='temperature')

cake_tidy

tastiness                                 
temperature            225        250        275        300
recipe position                                            
a      bottom    61.738655  53.912627  74.414730  98.786784
       top       51.709751  52.009735  68.576858  50.228470
b      bottom    57.095320  61.904369  61.196980  99.248541
       top       82.455004  95.224151  98.594881  58.169349
c      bottom    96.470207  52.001358  92.893227  65.473084
       top       71.306308  82.795477  92.098049  53.960273
d      bottom    52.799753  58.670419  51.747686  56.183110
       top       96.873178  76.101363  59.571620  50.971626

In [42]:
cake_tidy.reset_index(inplace=True)


cake_tidy

recipe position  tastiness                                 
temperature                        225        250        275        300
0                a   bottom  61.738655  53.912627  74.414730  98.786784
1                a      top  51.709751  52.009735  68.576858  50.228470
2                b   bottom  57.095320  61.904369  61.196980  99.248541
3                b      top  82.455004  95.224151  98.594881  58.169349
4                c   bottom  96.470207  52.001358  92.893227  65.473084
5                c      top  71.306308  82.795477  92.098049  53.960273
6                d   bottom  52.799753  58.670419  51.747686  56.183110
7                d      top  96.873178  76.101363  59.571620  50.971626

### Which recipe, on average, is the best? recipe b

In [38]:
cake_recipes.groupby('recipe').tastiness.mean()

recipe
a    63.922201
b    76.736074
c    75.874748
d    62.864844
Name: tastiness, dtype: float64

### Which oven temperature, on average, produces the best results? 275


In [46]:
cake_recipes.groupby('temperature').tastiness.mean()

temperature
225    71.306022
250    66.577437
275    74.886754
300    66.627655
Name: tastiness, dtype: float64

### Which combination of recipe, rack position, and temperature gives the best result? recipe b, bottom rack, 300 degrees


In [51]:
cake_recipes.sort_values(by = 'tastiness').tail(1)

temperature  tastiness recipe position
26         300  99.248541      b   bottom